cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/RunIds.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.app.runtime;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+import com.google.common.primitives.Longs;
+import org.apache.twill.api.RunId;
+
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.util.Enumeration;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Generates an unique ID for a running program using type 1 and variant 2 time-based {@link UUID}.
+ */
+public final class RunIds {
+  private static final Random RANDOM = new Random();
+
+  // Number of 100ns since 15 October 1582 00:00:000000000
+  private static final long NUM_100NS_INTERVALS_SINCE_UUID_EPOCH = 0x01b21dd213814000L;
+  // Multiplier to convert millisecond into 100ns
+  private static final long HUNDRED_NANO_MULTIPLIER = 10000;
+
+  private static final AtomicInteger COUNTER = new AtomicInteger();
+
+  /**
+   * @return UUID based on current time. If called repeatedly within the same millisecond, this is
+   * guaranteed to generate at least 10000 unique UUIDs for the millisecond.
+   */
+  public static RunId generate() {
```

Comment:
Or should I get the time from UUID and store it as the start time in MDS? This will eliminate the need to pass the start time into `ProgramRuntimeService`. What do you say?